src/main/cc/wfa/virtual_people/core/model/branch_node_impl.h

```diff
@@ -0,0 +1,116 @@
+#include "absl/status/statusor.h"
+#include "absl/strings/string_view.h"
+#include "absl/types/variant.h"
+#include "src/main/proto/wfa/virtual_people/common/model.pb.h"
+#include "wfa/virtual_people/core/model/model_node.h"
+#include "wfa/virtual_people/core/model/utils/distributed_consistent_hashing.h"
+#include "wfa/virtual_people/core/model/utils/field_filters_matcher.h"
+
+namespace wfa_virtual_people {
+
+// uint32_t: An index refers to a child node. Will be replaced with the
+//   corresponding std::unique_ptr<ModelNode> after calling
+//   ResolveChildReferences.
+// std::unique_ptr<ModelNode>: A child node.
+using ChildNodeRef = absl::variant<uint32_t, std::unique_ptr<ModelNode>>;
+
+// The implementation of the CompiledNode with branch_node set.
+//
+// The field branch_node in @node_config must be set.
+//
+// Currently selecting child node by chance or condition is supported.
+// TODO(@tcsnfkx): Implement attributes updater.
+// TODO(@tcsnfkx): Implement multiplicity.
+class BranchNodeImpl : public ModelNode {
+ public:
+  // Always use ModelNodeFactory.NewModelNode to get a ModelNode object.
+  // Users should never call the factory function or constructor of the derived
+  // class directly.
+  //
+  // Returns error status if any of the following happens:
+  //   @node_config.branch_node is not set.
+  //   @node_config.branch_node.branches is empty.
+  //   Neither node_index nor node is set in at least one of the
+  //     @node_config.branch_node.branches.
+  //   Neither chance nor condition is set in at least one of the
+  //     @node_config.branch_node.branches.
+  //   At least one of @node_config.branch_node.branches has chance set, and at
+  //     least one of @node_config.branch_node.branches has condition set.
+  static absl::StatusOr<std::unique_ptr<BranchNodeImpl>> Build(
+      const CompiledNode& node_config);
+
+  explicit BranchNodeImpl(
+      const CompiledNode& node_config,
+      std::vector<ChildNodeRef>&& child_nodes,
+      std::unique_ptr<DistributedConsistentHashing> hashing,
+      absl::string_view random_seed,
+      std::unique_ptr<FieldFiltersMatcher> matcher);
+  ~BranchNodeImpl() override {}
+
+  // Replaces all indexes in @child_nodes_ with the corresponding ModelNode
+  // objects.
+  // For any resolved index, the index / ModelNode object pair is deleted from
+  // @node_refs.
+  //
+  // Calling this method will resolve the child references for any node in the
+  // sub-tree starting from this node recursively.
+  //
+  // Returns error status if any of the following happens:
+  //   Any index in @child_nodes_ does not have an entry in @node_refs.
+  //   Neither index nor ModelNode is set in at least one entry in
+  //     @child_nodes_.
+  //   It returns error status when resolving the child references for any node
+  //     in the sub-tree.
+  //
+  // TODO(@tcsnfkx): Resolve the child references of the UpdateTree if exists.
+  absl::Status ResolveChildReferences(
```

Comment:
Should this be on ModelNode interface? We expect to only do it once starting from root node, correct?